### ISSUE TITLE
Sorting by group

### ISSUE THREAD

**Issue by jessp17** [royfrancis/pophelper#7]
Hi, I love this tool but am having a heck of a time sorting my samples by group.  I've been through tutorials, etc.   The individuals unfortunately are not in order by group (pop).  I'm reading in group labels via a text file (read.delim) and am using data from Admixture. 

I'm using the subsetgrp command to try to reorder and sort by the group ID.  However, I get this error:
"Error in pophelper:::grpLabels(df = df1, grplab = grplab[[pp]], subsetgrp = subsetgrp,  : 
  grpLabels: Duplicated contiguous block of grp labels. Not possible to subset or reorder grps.

Is this something I can work around?

Thanks

Thank you!

**Comment by royfrancis** [comment]
Have you tried adding this argument `ordergrp=TRUE`? If that doesn't work for you, feel free to get back to me.

**Comment by jessp17** [comment]
Hi Roy,

Well, I seemed to have missed that before.  In the mean time I went back to uploaded all new versions of R and (v3.4.0), and reinstalled the packages.

However, I now get an error when I try the plotQ with grplab=list(“lab1/n”=grp$V1),ordergrp=TRUE):
Error in verifyGrplab(grplab) :   verifyGrplab: Argument 'grplab' is not a data.frame object.

That was not an issue before.  grp is a data.frame and the dimensions are correct to fit with the Q datafiles.
It doesn’t seem to recognize grplab at all and I’m not sure where to go next.

I apologize for taking your time and also for asking a question previously asked – I just found that thread!

Thanks,
Jessica

**Comment by royfrancis** [comment]
In the current version of pophelper, `grplab` must be a `data.frame` rather than a `list`. This won't work: `grplab=list(“lab1/n”=grp$V1)`. This should work: `grplab=data.frame(lab1=grp$V1)`. Note that the `\n` is not needed anymore. I am sorry for these changes breaking existing code, but `data.frame` is easier when handling multiple label sets.

**Comment by jessp17** [comment]
Roy,
I now get this:

> plotQ(qlist=list[2],grplab=data.frame(lab1=grp$V1),ordergrp=TRUE)
Error in verifyGrplab(grplab) :  verifyGrplab: Argument 'grplab' contains one or more fields which are not character datatype.

I was wondering if my grp txt file is somehow messing things up but grplab seems to still be a problem.  My variable, grp, shows as.character = TRUE.

Jessica

**Comment by royfrancis** [comment]
Try `data.frame(lab1=grp$V1,stringsAsFactors=F)`.

**Comment by jessp17** [comment]
It’s still unhappy:

> plotQ(qlist=list[2],grplab=data.frame(lab1=grp$V1),stringsAsFactors=F)

Error in plotQ(qlist = list[2], grplab = data.frame(lab1 = grp$V1), stringsAsFactors = F) : unused argument (stringsAsFactors = F)

**Comment by royfrancis** [comment]
plotQ(qlist=list[2],grplab=data.frame(lab1=grp$V1,stringsAsFactors=F))
`stringsAsFactors=F` was in the wrong place.

**Comment by jessp17** [comment]
Yes – this works.  Thank you.  I greatly appreciate the time you put into this!

**Comment by mazepago** [comment]
Hello,

I was trying to visualize qplots of a range of different Ks, hoping that the color scheme could be more or less conservative across the graphs, as, for instance, CLUMPAK does on the same dataset: 
http://clumpak.tau.ac.il/results.html?jobId=1499954441

I tried to set the color palette with plotQ(clustercol), but this did not help in case of pophelper, althought that strategy seems to be efficient with CLUMPAK... 

Are conservative colors across Ks possible in principle with pophelper?

Best,
Glib

**Comment by royfrancis** [comment]
By conservative colours, you mean that clusters must have the same colours across runs and K? For that, the runs need to be reordered using CLUMPP.  Use `clumppExport(qlist=slist,useexe=T)` to export aligned files. Then use the aligned or merged files for plotting.

http://royfrancis.github.io/pophelper/#56_clumppexport

Colours are used in the same order from the palette. New clusters will take a new colour.

**Comment by mazepago** [comment]
Hi, i executed exactly that command, the only difference it is ADMIXTURE, not STRUCTURE list, but i doubted that would be the reason...

sketch of my commands:

## trial with just 3 iterations of 3 different Ks: 15-17

afiles<-list.files(path="/blabla/Iterations_01_03_Qs/", full.names=T)
alist <- readQ(files=afiles)
clumppExport(qlist=alist,useexe=T)
collectClumppOutput(filetype="both")
plotQ(qlist=readQ(afiles)[c(1:3)], imgoutput="join", clustercol=c("#1D72F5","#DF0101","#77CE61","#FF9326","#A945FF","#0089B2",
 "#FDF060","#FFA6B2","#BFF217","#60D5FD","#CC1577","#F2B950",
 "#7FB21D","#EC496F","#326397","#B26314","#027368","#A4A4A4", "#610B5E"))

The resulting figure is here: 
https://www.dropbox.com/s/qqda61sh0mo30az/Joined3Files-20170715100631.png?dl=0

the dropbox link to the both folders with inputs and outputs:
https://www.dropbox.com/sh/aa0fgomivor063a/AACyAmZBsrqz4AfRd0_x9104a?dl=0

Best,
G

**Comment by royfrancis** [comment]
Ah yes. So the feature to align cluster within each K is done by CLUMPP which has been incorporated into pophelper which is why that works. But the feature to align clusters across K is a feature of CLUMPAK. And that has not been incorporated into pophelper. So, aligning across K is not possible within pophelper. Sorry.

**Comment by KarlFetter** [comment]
Hi Roy,

I've got a bit of a puzzling problem with sorting populations by subsetgrp. I feel like I've done the due diligence of reading the manual, issues pages, and working with the package. I was even able to get populations sorted by my preference a few months ago for another project using an older version of pophelper (not sure which one), but now I'm using pophelper_2.2.7, ggplot2_3.0.0   Cairo_1.5-9.

The problem is that my populations are being sorted alphabetically, not according to the order that I input in subsetgrp. I can subset groups and plot fewer groups, but they always come out in alphabetical order. I ran some old code that worked from a few months back, and I got alphabetical order again, not the order I input.

I'll attach a screen shot of my session info and a bit of code. Perhaps there is a new interacting option with subsetgrp that I haven't set that needs to be set to override the alphabetical sorting??

What can I do to try and troubleshoot this?

Thanks for your help. Really prefer pophelper to my ggplot scripts, but I have to get it sorting again!


![screen shot 2018-09-07 at 11 10 10 am](https://user-images.githubusercontent.com/22380559/45227685-0fcd9980-b28f-11e8-9e43-2a1a42272d18.png)
